diff --git a/fs/f2fs/f2fs.h b/fs/f2fs/f2fs.h
index 4e696b65ed9b..91e1db24509f 100644
--- a/fs/f2fs/f2fs.h
+++ b/fs/f2fs/f2fs.h
@@ -3623,7 +3623,7 @@ static inline bool f2fs_force_buffered_io(struct inode *inode,
 	if (f2fs_encrypted_file(inode) &&
 	    !fscrypt_using_hardware_encryption(inode))
 		return true;
-	if (sbi->s_ndevs)
+	if (f2fs_is_multi_device(sbi))
 		return true;
 	/*
 	 * for blkzoned device, fallback direct IO to buffered IO, so
