diff --git a/fs/f2fs/file.c b/fs/f2fs/file.c
index f78e01439389..13ae8e0b798f 100644
--- a/fs/f2fs/file.c
+++ b/fs/f2fs/file.c
@@ -2578,7 +2578,7 @@ static int f2fs_ioc_flush_device(struct file *filp, unsigned long arg)
 							sizeof(range)))
 		return -EFAULT;
 
-	if (sbi->s_ndevs <= 1 || sbi->s_ndevs - 1 <= range.dev_num ||
+	if (!f2fs_is_multi_device(sbi) || sbi->s_ndevs - 1 <= range.dev_num ||
 			__is_large_section(sbi)) {
 		f2fs_msg(sbi->sb, KERN_WARNING,
 			"Can't flush %u in %d for segs_per_sec %u != 1\n",
