diff --git a/fs/f2fs/gc.c b/fs/f2fs/gc.c
index f38c4f514c14..1616f51ada31 100644
--- a/fs/f2fs/gc.c
+++ b/fs/f2fs/gc.c
@@ -1347,7 +1347,7 @@ void f2fs_build_gc_manager(struct f2fs_sb_info *sbi)
 	sbi->gc_pin_file_threshold = DEF_GC_FAILED_PINNED_FILES;
 
 	/* give warm/cold data area from slower device */
-	if (sbi->s_ndevs && !__is_large_section(sbi))
+	if (f2fs_is_multi_device(sbi) && !__is_large_section(sbi))
 		SIT_I(sbi)->last_victim[ALLOC_NEXT] =
 				GET_SEGNO(sbi, FDEV(0).end_blk) + 1;
 }
